### PR TITLE
[Feature:TAGrading] blind and unblind grading

### LINKED ISSUE #5349: Option for "Double-blind" Rubric Grading
**What problem are you trying to solve with Submitty**
In many cases it is sufficient and even desirable for rubric grading to be done anonymously, that is, without seeing the name of the student.  This can help remove or eliminate bias during exam grading, making sure the grader only grades the answer to the problem, and is not biased by knowing that student from class or office hours.

In most cases for peer grading, double blind grading is the correct default setting.  However, sometimes its important to only have single blind grading.  If peers evaluate in class project presentations, they will clearly know the identity of the student they are grading.

NOTE: Of course we cannot always guarantee anonymity of grading.  If the student writes their name at the top of every code file they submit.  Student names will be on the first page of their scanned exam pdf (and perhaps on every page of the exam).

Even if we primarily use double blind grading for a gradeable, it is often necessary for the instructor or full access grader to view the student's real name when following up on a specific grade inquiry, etc.

**Describe the way you'd like to solve this problem**

1.  Every gradeable with manual and/or peer rubric grading should have additional settings:

     manual grading: blind grading or unblinded grading [default]
     peer grading: double-blind [default] or single blind or unblinded

2.  The student index page for each gradeable (the page that current lists all submissions to grade with the student/teams real name) should be changed to use placeholder anonymous ids when the grading mode configuration is blind.

3. Instructors and full access graders (but not limited access or peer graders) should have a toggle option at the top of the student index page [next to randomize grading order] that lets them switch grading from the blind to unblinded, revealing all names.  This setting is stored in a cookie / local storage.  limited access graders and peer graders do not have this toggle.  The follow whatever was configured for this assignment.

4. Once the grader enters the grading page for a specific student, if the grading mode is blind, the black-bordered student information box is hidden.  And perhaps other select files (e.g., queue file?) are excluded from view.  Longer term we can also take steps to hide or "white out" student names from known / specified locations on scanned pdfs.

5. When students view their peer graded components, they are only show the names of the peer graders if the setting is "unblinded". 

Related to / replaces issues #4450 & #2661.

### PR DESCRIPTION
**Pull request by jchulton**
### Please check if the PR fulfills these requirements:

* [ ] The PR title and message follows our [guidelines](https://submitty.org/developer/how_to_contribute)
* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant
closes #5349 
### What is the current behavior?
All peer grading is double blind grading (neither the student nor the grader know eachother's id). All other grading is non-blind (the grader knows the student id) .
### What is the new behavior?
Peer grading for a gradeable can be set to double blind, single blind (the grader can see the student id, but the student can't see the grader id) or unblind in the edit gradeable menu. Limited Access grader grading can be set to unblind or blind. Instructors and full access graders can enter an anon grading mode in the grading details page (similar to blind grading).

image

### CODE REVIEW

**Comment by bmcutler** [comment]
Got this error attempting to either grade or open create edit gradeable on an existing gradeable:

Oh no! Something irrecoverable has happened...

ParseError (Code: 0) thrown in /usr/local/submitty/site/app/views/grading/ElectronicGraderView.php (Line 850) by:
        if (!$peer && $gradeable->getLimitedAccessBlind() !== 2l) {


Message:
syntax error, unexpected 'l' (T_STRING)

Stack Trace:
#0 /usr/local/submitty/site/vendor/composer/ClassLoader.php(322): Composer\Autoload\includeFile('/usr/local/submitty/site/vendor/composer/../../app/views/grading/ElectronicGraderView.php')
#1 unknown file(unknown line): Composer\Autoload\ClassLoader->loadClass('app\views\grading\ElectronicGraderView')
#2 /usr/local/submitty/site/app/libraries/Output.php(374): spl_autoload_call('app\views\grading\ElectronicGraderView')
#3 /usr/local/submitty/site/app/libraries/Output.php(201): app\libraries\Output->getView('app\views\grading\ElectronicGraderView')
#4 /usr/local/submitty/site/app/libraries/Output.php(170): app\libraries\Output->renderTemplate('grading\ElectronicGrader', 'popupStudents')
#5 /usr/local/submitty/site/app/controllers/admin/AdminGradeableController.php(268): app\libraries\Output->renderOutput(Array, 'popupStudents')
#6 /usr/local/submitty/site/app/controllers/admin/AdminGradeableController.php(28): app\controllers\admin\AdminGradeableController->editPage(app\models\gradeable\Gradeable, 'f20', 'sample', 0)
#7 unknown file(unknown line): app\controllers\admin\AdminGradeableController->editGradeableRequest('peer_gradeable', 0)
#8 /usr/local/submitty/site/app/libraries/routers/WebRouter.php(217): call_user_func_array(Array, Array)
#9 /usr/local/submitty/site/app/libraries/routers/WebRouter.php(195): app\libraries\routers\WebRouter->run()
#10 /usr/local/submitty/site/public/index.php(127): app\libraries\routers\WebRouter::getWebResponse(Symfony\Component\HttpFoundation\Request, app\libraries\Core)



============================


Got this error when attempting to make a new gradeable:




Oh no! Something irrecoverable has happened...

app\exceptions\DatabaseException (Code: 0) thrown in /usr/local/submitty/site/app/libraries/database/DatabaseQueries.php (Line 4192) by:
                throw new DatabaseException("Electronic gradeable didn't have an entry in the electronic_gradeable table!");


Message:
Electronic gradeable didn't have an entry in the electronic_gradeable table!

Stack Trace:
#0 unknown file(unknown line): app\libraries\database\DatabaseQueries->app\libraries\database\{closure}(Array)
#1 /usr/local/submitty/site/app/libraries/database/DatabaseRowIterator.php(56): call_user_func(Closure, Array)
#2 /usr/local/submitty/site/app/models/gradeable/GradeableList.php(72): app\libraries\database\DatabaseRowIterator->next()
#3 /usr/local/submitty/site/app/controllers/NavigationController.php(29): app\models\gradeable\GradeableList->__construct(app\libraries\Core)
#4 unknown file(unknown line): app\controllers\NavigationController->navigationPage()
#5 /usr/local/submitty/site/app/libraries/routers/WebRouter.php(217): call_user_func_array(Array, Array)
#6 /usr/local/submitty/site/app/libraries/routers/WebRouter.php(195): app\libraries\routers\WebRouter->run()
#7 /usr/local/submitty/site/public/index.php(127): app\libraries\routers\WebRouter::getWebResponse(Symfony\Component\HttpFoundation\Request, app\libraries\Core)

**Comment by bmcutler** [comment]
@jchulton 
Please merge master and hook up/tweak the double-blind/single-blind variables in 
site/app/templates/grading/electronic/Status.twig
as appropriate.
Thanks!

**Comment by jchulton** [comment]
The variables have been hooked up

File: site/app/templates/grading/electronic/Status.twig [hunk 1]
```diff
@@ -200,6 +197,11 @@
 
                 Grading Index
             </a>
+            
+            <fieldset>
+
+        </fieldset>
```
Comment (emaicus) [inline site/app/templates/grading/electronic/Status.twig:203]: Why the empty fieldset?

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 2]
```diff
@@ -1079,10 +1079,12 @@ public function adminTeamSubmit($gradeable_id) {
      * @Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/grading/grade")
      * @Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/grading/grade/{version}", requirements={"version"="^beta$"})
      */
-    public function showGrading($gradeable_id, $who_id = '', $from = "", $to = null, $gradeable_version = null, $sort = "id", $direction = "ASC", $to_ungraded = null, $component_id = "-1", $version = null) {
-        // If the version is set, user is requesting for new TA grading interface
+    public function showGrading($gradeable_id, $who_id = '', $from = "", $to = null, $gradeable_version = null, $sort = "id", $direction = "ASC", $to_ungraded = null, $component_id = "-1", $anon_mode = false, $version = null) {
         $showNewInterface = isset($version);
-
+        $anon_mode = false;
```
[the remark this answers is not in the document]
Comment (jchulton) [reply site/app/controllers/grading/ElectronicGraderController.php]: Removed setting anon_mode to false. Tested and seems fine.

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 3]
```diff
@@ -1092,6 +1094,20 @@ public function showGrading($gradeable_id, $who_id = '', $from = "", $to = null,
         }
         $peer = $gradeable->isPeerGrading() && $this->core->getUser()->getGroup() == User::GROUP_STUDENT;
         $team = $gradeable->isTeamAssignment();
+        if ($gradeable->isPeerGrading() && $this->core->getUser()->getGroup() == User::GROUP_STUDENT) {
+            $peer = true;
+        }
+        
+        $peer_double_blind = false;
+        if ($peer && $gradeable->getPeerBlind() == 3) {
```
[the remark this answers is not in the document]
Comment (jchulton) [reply site/app/controllers/grading/ElectronicGraderController.php]: Removed this line

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 4]
```diff
@@ -1092,6 +1094,20 @@ public function showGrading($gradeable_id, $who_id = '', $from = "", $to = null,
...
+            $peer_double_blind = true;
+        }
+        
+        $limited_access_blind = false;
+        
+        if ($gradeable->getLimitedAccessBlind() == 2 && $this->core->getUser()->getGroup() == User::GROUP_LIMITED_ACCESS_GRADER) {
```
[the remark this answers is not in the document]
Comment (jchulton) [reply site/app/controllers/grading/ElectronicGraderController.php]: Removed this line

File: site/app/templates/grading/electronic/Status.twig (hunk 1 above)
Comment (jchulton) [reply site/app/templates/grading/electronic/Status.twig]: Removed

File: site/app/controllers/grading/ElectronicGraderController.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (jchulton) [reply site/app/controllers/grading/ElectronicGraderController.php]: I've added a enumeration to Gradeable.php, similar to the user groups in User.php

**Comment by jchulton** [comment]
I've added an amIBlindGrading function and enum to Gradeable.php for the blind grading. Hopefully this makes things clearer.

File: tests/e2e/base_testcase.py [hunk 5]
```diff
@@ -9,7 +9,7 @@
 from urllib.parse import urlparse
 
 from selenium import webdriver
-from websocket import create_connection
+#from websocket import create_connection
```
Comment (emaicus) [inline tests/e2e/base_testcase.py:12]: Un-comment this to help fix travis

Comment (jchulton) [reply tests/e2e/base_testcase.py]: Fixed

**Comment by bmcutler** [comment]
@jchulton @emaicus 
What is the status of this PR?

**Comment by jchulton** [comment]
I retested it and it seems to work fine.

**Review by emaicus — CHANGES_REQUESTED** [review]
This PR is mostly working, but isn't perfect. Student ids are being exposed in the source of the grading and submission pages (after grades are released).

**Comment by emaicus** [comment]
I advise that this PR be updated to pass Travis, then merged. Once this PR is merged, work should be done on #5783 to fix any new issues. I can then review #5783 so that it can be merged.

**Comment by jchulton** [comment]
Alright, this PR should be good to merge then

**Comment by jchulton** [comment]
I believe I have patched all of the student id exposure (on a separate branch but I can merge to this one). The testing PR is proving more difficult due to changes to sample courses.

**Pull request merged**